source/channel_bootstrap.c

```diff
@@ -434,6 +692,112 @@ struct server_channel_data {
     struct server_connection_args *server_connection_args;
 };
 
+static void s_tls_server_on_negotiation_result(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    int err_code,
+    void *user_data) {
+    struct server_connection_args *connection_args = user_data;
+
+    if (connection_args->user_on_negotiation_result) {
+        connection_args->user_on_negotiation_result(handler, slot, err_code, connection_args->tls_user_data);
+    }
+
+    if (!err_code) {
+        connection_args->incoming_callback(
+            connection_args->bootstrap, AWS_OP_SUCCESS, slot->channel, connection_args->user_data);
+    } else {
+        /* the tls handler is responsible for calling shutdown when this fails. */
+        connection_args->shutdown_callback(connection_args->bootstrap, err_code, NULL, connection_args->user_data);
+    }
+}
+
+/* in the context of a channel bootstrap, we don't care about these, but since we're hooking into these APIs we have to
+ * provide a proxy for the user actually receiving their callbacks. */
+static void s_tls_server_on_data_read(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    struct aws_byte_buf *buffer,
+    void *user_data) {
+    struct server_connection_args *connection_args = user_data;
+
+    if (connection_args->user_on_data_read) {
+        connection_args->user_on_data_read(handler, slot, buffer, connection_args->tls_user_data);
+    }
+}
+
+/* in the context of a channel bootstrap, we don't care about these, but since we're hooking into these APIs we have to
+ * provide a proxy for the user actually receiving their callbacks. */
+static void s_tls_server_on_error(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    int err,
+    const char *message,
+    void *user_data) {
+    struct server_connection_args *connection_args = user_data;
+
+    if (connection_args->user_on_error) {
+        connection_args->user_on_error(handler, slot, err, message, connection_args->tls_user_data);
+    }
+}
+
+static inline int s_setup_server_tls(struct server_connection_args *connection_args, struct aws_channel *channel) {
+    struct aws_channel_slot *tls_slot = NULL;
+    struct aws_channel_handler *tls_handler = NULL;
+
+    /* as far as cleanup goes here, since we're adding things to a channel, if a slot is ever successfully
```

Comment:
Same comment here as above, I worry about the asymmetrical/non-atomic nature of these functions sometimes not cleaning up their broken state/resources, and there's no indication what the distinction is between when this function manages resources vs they're left to be effectively garbage collected.

Reply:
yeah, agreed, we need to clean those semantics up, in a different PR